cudaPackages: introduce and use cudaLib

ConnorBaker opened the pull request:
Broadly, this PR does these things:

- Introduces a new top-level attribute, `_cuda`, which consolidates CUDA package set backbone (static data, configurations, utility functions, etc. required to create a CUDA package set)
- Removes things from the CUDA package set fixed point which should be supplied by `callPackage` and the `newScope` used to construct the package set (i.e., `_cuda.bootstrapData`, `_cuda.fixups`)
- Aligns language regarding redistributables: `redist arch` has been changed to `redist system` to better mirror `nix system`, as both are two-tuple system configs
- Rewrites `cudaPackages.backendStdenv`

This PR is a step toward landing more improvements from https://github.com/ConnorBaker/cuda-packages and merging the changes in #406740.

This PR presents largely organizational changes, but does cause rebuilds due to changes in the default selection of CUDA capabilities (see the section under `_cuda.lib`).

The `_cuda` attribute and all that it contains is introduced to enable re-use out of tree. For the time being, it is provided without any guarantee of stability.

### `_cuda.fixups`

- `fixups` is moved to `_cuda.fixups`
- `_cuda.fixups` is removed from CUDA package sets and exists only at top-level to make clear it exists outside of any CUDA package set's fixed point

### `_cuda.bootstrapData`

- GPU capabilities are rewritten to include more information, updated with new architectures and feature sets, and brought under `_cuda.bootstrapData.cudaCapabilityToInfo` (`pkgs/development/cuda-modules/_cuda/db/bootstrap/cuda.nix`)
- NVCC compatibilities are rewritten to group versions for host compilers under an attribute set rather than using a string prefix and brought under `_cuda.bootstrapData.nvccCompatibilities` (`pkgs/development/cuda-modules/_cuda/db/bootstrap/nvcc.nix`)
- Additional commonly used data are available (e.g., `redistNames`, `redistUrlPrefix`)

### `_cuda.lib`

Constructing out-of-tree package sets for CUDA packages is difficult and typically requires vendoring large numbers of files to be able to benefit from the machinery we (the CUDA team) have built. `_cuda.lib` provides an interface through which out-of-tree consumers of Nixpkgs can access the same machinery we use. (Data and configuration re-use are enabled by `_cuda.bootstrapData` and `_cuda.fixups`.)

`_cuda.lib` also serves to centralize and document a number of ad-hoc functionalities created in the CUDA package set. For example, many of the functions which had been lumped into `flags.nix` are now available in `_cuda.lib`.

Due to its recency, very little of `_cuda.lib` is provided as a stable API. Many functions are only used internally at the moment (they are prefixed with an `_`) and are provided without guarantees of stability or correctness. They are however, available for use given their utility.

#### Changes to the default set of CUDA capabilities

Previously, the set of default capabilities for CUDA 12.8 were

```
6.0 6.1 7.0 7.5 8.0 8.6 8.9 9.0 9.0a 10.0 12.0
```

They are now

```
            7.5 8.0 8.6 8.9 9.0      10.0 12.0
```

Given the number of capabilities introduced by the last few releases, dropping older capabilities becomes a necessity to keep the amount of device code generated small enough to link against. (There is a 2GB limit.) Since recent TensorRT releases have deprecated and no dropped support for `6.0` through `7.0`, it makes sense to drop them from the default set of capabilities.

Additionally, the inclusion of architecture-specific capabilities, like `9.0a`, have been removed from the default collection of targets, as they are not meant to be used with other capabilities without careful guarding in source (a feature only discussed in 12.9[^1]).

### `cudaPackages.backendStdenv`

The changes made to `backendStdenv` enable evaluation-time checking of certain common mistakes. Additionally, it provides a single point of access to common attributes describing the CUDA package set, such as:

- `nvccHostCCMatchesStdenvCC`: whether the current `stdenv` matches the version which will be provided to NVCC
- `hostNixSystem`: an alias to `stdenv.hostPlatform.system`
- `hostRedistSystem`: the NVIDIA equivalent of `hostNixSystem`
- `cudaCapabilities`: the resolved collection of CUDA capabilities `cudaPackages.flags`'s attributes target
- `hasJetsonCudaCapability`: whether `cudaCapabilities` contains a capability belonging to a Jetson device

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

[^1]: https://developer.nvidia.com/blog/nvidia-blackwell-and-nvidia-cuda-12-9-introduce-family-specific-architecture-features/#architecture-specific_features

ConnorBaker committed:
cudaLib: init

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages: switch to cudaLib

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages: rewrite backendStdenv

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages: add cudaNamePrefix

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages: doc fixup

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.driver_assistant: mark as unsupported

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 406531`

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>nixpkgs-manual</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>nixpkgs-manual</li>
  </ul>
</details>

ConnorBaker commented:
Changes:

- rebased on master after #406545 was merged
- added `cudaLib.utils.allowUnfreeCudaPredicate`
- added commit for tree-wide changes to update usage of `cudaPackages.flags`

SomeoneSerge commented on pkgs/top-level/all-packages.nix:
```diff
@@ -2732,6 +2732,8 @@ with pkgs;
 
   cron = isc-cron;
 
+  cudaLib = import ../development/cuda-modules/lib;
+
   cudaPackages_11_0 = callPackage ./cuda-packages.nix { cudaMajorMinorVersion = "11.0"; };
   cudaPackages_11_1 = callPackage ./cuda-packages.nix { cudaMajorMinorVersion = "11.1"; };
```
Here - instead of exposing `cudaLib` I see something like `makeCudaScope = let cudaLib = ...; in cudaLib.${...}; cudaPackages_XX_YY = makeCudaScope { ... }`

ConnorBaker replied:
I would like to have `cudaLib` exposed outside of the CUDA package set since it does not depend on the CUDA version or architecture. Having it at the top level is also important because if out-of-tree consumers modify it (e.g., by adding new GPUs), those changes are propagated to all consumers of `cudaLib`. One such example where that's important is the `pkgsCuda` attribute set I'd like to introduce in https://github.com/NixOS/nixpkgs/pull/406568: https://github.com/NixOS/nixpkgs/pull/406568/commits/eb044ff0e142cfe5cef8f57d0d1557ac4d1bb253.

SomeoneSerge commented:
There are three separate things here: data, configuration, and helper
functions. I'd like to have public the data (I should push a draft today
still...) and a single function: `DB -> Configuration -> PackageSet`. Wdyt?


On 12/05/2025 21.09, Connor Baker wrote:
> @ConnorBaker commented on this pull request.
>
>
>
>> +  cudaLib = import ../development/cuda-modules/lib;
> +
>    cudaPackages_11_0 = callPackage ./cuda-packages.nix { cudaMajorMinorVersion = "11.0"; };
>    cudaPackages_11_1 = callPackage ./cuda-packages.nix { cudaMajorMinorVersion = "11.1"; };
>
> I would like to have `cudaLib` exposed outside of the CUDA package set since it does not depend on the CUDA version or architecture. Having it at the top level is also important because if out-of-tree consumers modify it (e.g., by adding new GPUs), those changes are propagated to all consumers of `cudaLib`. One such example where that's important is the `pkgsCuda` attribute set I'd like to introduce in https://github.com/NixOS/nixpkgs/pull/406568: https://github.com/NixOS/nixpkgs/pull/406568/commits/eb044ff0e142cfe5cef8f57d0d1557ac4d1bb253.
>
-- 
SomeoneSerge <https://someonex.net>

ConnorBaker commented:
> There are three separate things here: data, configuration, and helper functions. I'd like to have public the data (I should push a draft today still...) and a single function: `DB -> Configuration -> PackageSet`. Wdyt?

So long as the data and configuration are reusable in such a way that out-of-tree consumers who want to change or make new CUDA package sets can use or modify them without copy-pasting a bunch of files from Nixpkgs to their local tree, that sounds good to me!

ConnorBaker commented on pkgs/development/cuda-modules/tests/flags.nix:
```diff
@@ -0,0 +1,78 @@
+        "sm_86"
+      ];
+
+      virtualArchs = [
```
Since `architecture` becomes `arch`, the plural `architectures` should become `archs`.

ConnorBaker replied:
I'll change it back :)

ConnorBaker commented:
Rebased on master and introduced no-warn aliases to `cudaPackages.flags` to avoid breaking out-of-tree users. These will be moved to `pkgs/development/cuda-modules/aliases.nix` after the release branch-off.

SomeoneSerge reviewed (commented):
Awesome, thanks! Overall, want to merge asap. The only parts I felt any opposition to are `Archs` and the exposure of the complete lib as public api. Synchronization-wise, would be nice to come up with something regarding the `data.nix` comment.

I only briefly glanced at `cuda-packages.nix`, it's getting late on this hemisphere...

ConnorBaker commented on doc/languages-frameworks/cuda.section.md:
```diff
@@ -115,8 +115,8 @@ All new projects should use the CUDA redistributables available in [`cudaPackage
-1. Update `nvcc-compatibilities.nix` in `pkgs/development/cuda-modules/` to include the newest release of NVCC, as well as any newly supported host compilers.
-2. Update `gpus.nix` in `pkgs/development/cuda-modules/` to include any new GPUs supported by the new release of CUDA.
+1. Update `nvccCompatibilities` in `pkgs/development/cuda-modules/lib/data.nix` to include the newest release of NVCC, as well as any newly supported host compilers.
+2. Update `cudaCapabilityToInfo` in `pkgs/development/cuda-modules/lib/data.nix` to include any new GPUs supported by the new release of CUDA.
```
It was, but I wasn't fond of it because it wasn't descriptive -- I like `cudaCapabilityToInfo` because it makes it clear that it's an attribute set (or a function which implements a mapping) from a CUDA capability to some information pertaining to that capability (when I was using the module system in http://github.com/connorbaker/cuda-packages, I actually had a `cudaCapabilityInfo` type).

So, really `cudaCapabilityToInfo` is the short version of `cudaCapabilityToCudaCapabilityInfo` :)

ConnorBaker commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -17,7 +17,7 @@
   # You can find a full list here: https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/
   # For example if you're on an RTX 3060 that means you're using "Ampere" and you need to pass "sm_86"
-  cudaArches ? cudaPackages.flags.realArches or [ ],
+  cudaArches ? cudaPackages.flags.realArchs or [ ],
```
Will do

ConnorBaker commented on pkgs/development/cuda-modules/cusparselt/extension.nix:
```diff
@@ -54,17 +50,12 @@ let
     (attrsets.attrByPath [
       pname
-      redistArch
+      redistSystem
```
Aww, you think that just because their components for system configuration are backwards relative to Nix, we should give it a backwards name?

I'm obviously partial to `redistSystem` and I like that it mirrors `nixSystem`. I also like that I wouldn't have to find and replace all occurrences of it if we keep it :)

SomeoneSerge replied:
But `nixSystem` is also something we only have in `cuda-modules/`. My logic behind the `nvidia` suffix is that it's used for the runfile too, and it serves the same function as Nixpkgs' "system", just written in a different style (nvidia style?)

ConnorBaker replied:
`nixSystem` refers (through some number of steps) to the `system` (or `hostPlatform.system`) value given when importing Nixpkgs -- I chose that name to make it clear that it's the `system` value (`stdenv.hostPlatform.system`) coming from Nix/Nixpkgs.

I guess I mean to say I'd really prefer with stick with `System` as a suffix -- it makes sense in my head since `nix` and `redist` in `nixSystem` and `redistSystem` both function as qualifying nouns, modifying `System`. Put another way, I read `systemNvidia` in the same uncomfortable way I would read `soup chicken` -- it feels backwards.

I'd be okay with using `nvidiaSystem` instead of `redistSystem`, as that's probably more descriptive.

SomeoneSerge replied:
> nixSystem refers (through some number of steps) to the system (or hostPlatform.system) value given when importing Nixpkgs -- I chose that name to make it clear that it's the system value (stdenv.hostPlatform.system) coming from Nix/Nixpkgs.

My feeling is this is the default in Nixpkgs, so I'd not spawn additional names?

SomeoneSerge replied:
Like, the name `nixSystem` actually made me doubt whether it's actually `system` or something completely unrelated

ConnorBaker commented on pkgs/development/cuda-modules/lib/data.nix:
````diff
@@ -0,0 +1,690 @@
+    redistSystems :: [String]
+    ```
+  */
+  redistSystems = [
+    "linux-aarch64"
+    "linux-all" # Taken to mean all other linux systems
+    "linux-sbsa"
+    "linux-x86_64"
+    "source" # Source-agnostic platform
````
> Parts like this (or, say, `redistNames`) can be derived from JSON, unlike e.g. the compatibility tables or the gencode lists.

Yes, things like `redistSystems` could be derived, but some of them I'm not so sure about. For example, TensorRT -- there's still no redistrib manifest for it, so I resorted to making my own: https://github.com/ConnorBaker/cuda-packages/tree/main/pkgs/development/cuda-modules/manifests/tensorrt. How would we handle things which can be *mostly* derived (like `redistNames`)?

> If we merge the db PR after cudaLib, I'd like to keep this file in place and include into the "db" without modification (except for removing the derivable bits maybe). Is there a name we could give this, other than `data.nix`, suggestive of it being the "human-maintained knowledge" as opposed to the derived/code-generated stuff?

Hmm... I personally would put derived data in a `derived.nix` file.

I get what you're saying about the name `data.nix` -- it's certainly not clear from the name alone that it is "human-maintained knowledge" which needs to be kept up to date... and... tended to? Like a garden? It's not the sole thing we'd need to create the package set, so I'd be hesitant to call it `seed.nix`... hell we could split the `nvccCompatibilities` and `cudaCapabilityToInfo` into their own files under a `data` directory and just name them as such (so the current `data.nix` becomes `data/default.nix`).

SomeoneSerge replied:
> Yes, things like redistSystems could be derived, but some of them I'm not so sure about. For example, TensorRT -- there's still no redistrib manifest for it,

Well that's why the derived stuff is extended with the "human knowledge" (or the other way around if you prefer). Regarding platforms, at the moment the derived list doesn't seem to miss any (it's a superset because of ppc and windows stuff): https://cuda-index.someonex.net/sidx?sql=SELECT%0D%0A++DISTINCT+plat.str+AS+platform%0D%0AFROM%0D%0A++CudaArtifact+AS+ca%0D%0A++LEFT+JOIN+Str+AS+plat+ON+ca.platform+%3D+plat.id (also https://cuda-index.someonex.net/sidx/cuda_platforms for where they come from)

SomeoneSerge replied:
> I personally would put derived data in a derived.nix file.

Yes, I considered, I like this: we could have entry point at `data.nix` (`data/default.nix`), and then separately `schema.nix` and `derived.nix`. The real reason I posted the comment is I'm afraid I'll end up changing the indents in this file and screwing up the git-blame yet again. I reeeeeally want to get to a place where I can start relying on git-blame again but we keep doing global changes

ConnorBaker commented on pkgs/development/cuda-modules/packages/backendStdenv.nix:
```diff
@@ -1,32 +1,154 @@
+    ];
 
-lib.extendDerivation assertCondition passthruExtra cudaStdenv
+  assertCondition = addErrorContext "while evaluating ${mkVersionedName "cudaPackages" cudaMajorMinorVersion}.backendStdenv" (
+    _evaluateAssertions assertions
```
Thank you! It's a result of getting mega-frustrated with evaluation errors lmao

ConnorBaker commented on pkgs/development/cuda-modules/packages/backendStdenv.nix:
```diff
@@ -1,32 +1,154 @@
-# This is what nvcc uses as a backend,
-# and it has to be an officially supported one (e.g. gcc11 for cuda11).
-#
-# It, however, propagates current stdenv's libstdc++ to avoid "GLIBCXX_* not found errors"
-# when linked with other C++ libraries.
-# E.g. for cudaPackages_11_8 we use gcc11 with gcc12's libstdc++
```
I felt the comment block should be attached to the actual definition of `backendStdenv` since the file is longer now

SomeoneSerge replied:
counter-argument: at the top the reader discovers what's this for as soon as the open the file; I'd maybe even go the opposite way, and add a `::` type signature of the "return" value of the file at the very top before the comment

ConnorBaker replied:
Moved back to the top of the file in the latest

ConnorBaker commented:
> Awesome, thanks! Overall, want to merge asap. The only parts I felt any opposition to are `Archs` and the exposure of the complete lib as public api.

Alright, you've won me over on `arches` vs `archs` -- I'll change it back to `archs` when I get a chance.

As for the library, are the leading underscores and added comments not enough? I'm okay with making more of it private (in the same way as before, with underscores and comments) -- I really just want to be able to re-use it immediately.

> Synchronization-wise, would be nice to come up with something regarding the `data.nix` comment.

Agreed! Brain is all out, so I'll have to come back around to this.

> I only briefly glanced at `cuda-packages.nix`, it's getting late on this hemisphere...

Sleep is good!

SomeoneSerge commented on pkgs/development/cuda-modules/lib/default.nix:
> What's the next suggestion, 📚?

Interesting point, how do I add that to my keyboard?

SomeoneSerge commented on pkgs/top-level/cuda-packages.nix:
```diff
@@ -37,41 +38,35 @@ let
 
   passthruFunction = final: {
     inherit
+      cudaLib
```
Spotted! I saw some occurrences of `final.cudalib`, but I it was inside the library, referring to the `makeExtensible` thing.

This bit, I believe, is very important: `cudaPackages` should not have their own `cudaLib` fixpoints because then they can have different (diverging) "libraries"; in the current version of the PR, `callPackage` should be picking `cudaLib` from the top-level! I think it's important that the "library" (esp. the data) be not affected by the "configuration" of any particular `cudaPackages` set. The layering should be: `configuration` `->` (identifies/is part of) `cudaPackages` `->` (is built using) `cudaLib`. No arrows should point in the reverse directions

SomeoneSerge replied:
So all `cudaLib` knows about "package sets" and "configurations" is it implements functions for interpreting configurations and resolving them against data

ConnorBaker replied:
I like how you phrased that -- in the latest I added a comment in `pkgs/top-level/cuda-packages.nix` which says about as much.

ConnorBaker committed:
tree-wide: cudaPackages.flags updates

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.fixups -> pkgs.cudaFixups

SomeoneSerge commented on pkgs/development/cuda-modules/lib/default.nix:
```diff
@@ -0,0 +1,13 @@
+let
+  lib = import ../../../../lib;
+in
+lib.fixedPoints.makeExtensible (final: {
```
Side note: do I understand correctly that with auto-called args the `import ./default.nix { }` would always be a think even though the function object itself is in the import cache?

ConnorBaker replied:
I'm afraid I don't understand with respect to "auto-called args" and the snippet you have with `import` -- are you talking about `callPackage`?

SomeoneSerge replied:
I meant the top-level lambda args ("auto-called" as in nix-build/nix-shell applying `{}` to lambdas).

Basically, the question is, do we benefit from import cache in case of `{ foobar ? ... }:`, or only in case of `let foobar = ...; in`?

ConnorBaker commented:
Rebased on master, changed `archs` back to `arches`, split up the directory structure of `cudaLib` to group functions semantically and make more clear what's being exported.

Also moved `cudaLib`, `cudaFixups` (was `fixups`) to the top level and removed them from the package fixed-point, along with `lib` and `pkgs` (which were both available through `newscope` anyway).

ConnorBaker commented on doc/languages-frameworks/cuda.section.md:
```diff
@@ -115,8 +115,8 @@ All new projects should use the CUDA redistributables available in [`cudaPackage
 
-1. Update `nvcc-compatibilities.nix` in `pkgs/development/cuda-modules/` to include the newest release of NVCC, as well as any newly supported host compilers.
-2. Update `gpus.nix` in `pkgs/development/cuda-modules/` to include any new GPUs supported by the new release of CUDA.
+1. Update `nvccCompatibilities` in `pkgs/development/cuda-modules/lib/data.nix` to include the newest release of NVCC, as well as any newly supported host compilers.
```
See changes in latest

SomeoneSerge commented:
Tagged for backport, the PR should only contain backwards-compatible changes: we reproduce the original public interface (essentially, top-level derivations exposed by `cudaPackages`, and most of `cudaPackages.flags` - although we're yet to define the subset precisely) using slightly different internals. The intent is we'll have exposed the experimental out-of-tree customization support in the release branch (opt-in and no liability), and ensure smoother backporting of any potential future PRs.

My current assessment is

1. we can rename the top-level fixpoint from `cudaLib` to something like `_cudaInternals`,
2. we may backport "breaking" changes to `_cudaInternals` (only `cudaPackages` is public)
3. we keep the static/seed data in `_cudaInternals`,
4. we instantiate the "db" (https://github.com/NixOS/nixpkgs/pull/406740) inside "internals", passing it the seed data,
5. (layering/arrow directions) we assume the convention that
   1. the "seed data" inside the fixpoint must not have any dependencies,
   2. the "db" may only depends on "seed data",
   3. a cuda package set may depend on the db but may not "own" (instantiate) a db,
   4. a cuda package set may depend on a "package set configuration", which
      0. requests host and target platforms,
      1. requests capabilities,
      2. maps "pname" to an optional "version"
      3. maps "compatibility tag names" to optional "compatibility tags" (e.g. "cuda" to "cuda12")
   5. `cudaLib` part of `_cudaInternals` may include a function generating a "cuda package set configuration" from
     1. the db
     2. `config`
     3. `{host,target}Platform`
     4. a structure mapping "product name" (e.g. "cuda", "cudnn") to "product release" ("product version"; e.g. `"cuda" => "12.8", "cudnn" => "8.9"`); the pair ("product name", "product version") must be registered in the db; the mapping of `pname` to `version` is completed with `null`s if necessary

Suggestions, objections?

ConnorBaker committed:
_cuda: introduce to organize CUDA package set backbone

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker commented:
Okay! Per our conversation earlier today @SomeoneSerge:

- All of `cudaLib`, `cudaData`, and `cudaFixups` exist in a single fixed point outside of all the CUDA package set fixed points, at the top-level inside the attribute `_cuda` (as an alternative to `_cudaInternal`)
- `cudaLib` is now just the utilities (it's what was previously called `cudaLib.utils`)
- All the data is split out into `cudaData`
- `fixups` is now `cudaFixups`

Your changes in #406740 should now be able to be added to, or consolidated with, `cudaData`.

SomeoneSerge commented on pkgs/development/cuda-modules/_cuda/data/cuda.nix:
````diff
@@ -0,0 +1,355 @@
+    ];
+    ```
+  */
+  allSortedCudaCapabilities = lib.sort lib.versionOlder (lib.attrNames cudaData.cudaCapabilityToInfo);
````
Just one last bit: let's move `_cuda/data` to something like `_cuda/static-data`, otherwise it'll be a mass rename in the next PR.

I'll then move `allSortedCudaCapabilities` to "db" and remove `cudaData` from the arguments here, so that the "static" bit is free of fixpoints

SomeoneSerge commented on pkgs/top-level/all-packages.nix:
```diff
@@ -2726,6 +2726,21 @@ with pkgs;
+  # set. To avoid this, we declare `_cuda` and inherit its members here, at top-level. (This also allows us to benefit
+  # from import caching, as it should be evaluated once per system, rather than per-system and CUDA package set.)
+  _cuda = import ../development/cuda-modules/_cuda;
+  inherit (_cuda) cudaData cudaLib cudaFixups;
```
These we should keep private.

SomeoneSerge commented on pkgs/top-level/all-packages.nix:
```diff
@@ -2726,6 +2726,21 @@ with pkgs;
 
   cron = isc-cron;
 
+  # The _cuda attribute set is a fixed-point which contains the static functionality required to construct CUDA package
+  # sets. For example, `_cuda.cudaData` includes information about NVIDIA's redistributables (such as the names NVIDIA
+  # uses for different systems), `_cuda.cudaLib` contains utility functions like `formatCapabilities` (which generate
+  # common arguments passed to NVCC and `cmakeFlags`), and `_cuda.cudaFixups` contains `callPackage`-able functions
+  # which are provided to the corresponding package's `overrideAttrs` attribute to provide package-specific fixups
+  # out of scope of the generic redistributable builder.
+  #
+  # Since this attribute set is used to construct the CUDA package sets, it must exist outside the fixed point of the
+  # package sets. Make these attributes available directly in the package set construction could cause confusion if
+  # users override the attribute set with the expection that changes will be reflected in the enclosing CUDA package
+  # set. To avoid this, we declare `_cuda` and inherit its members here, at top-level. (This also allows us to benefit
+  # from import caching, as it should be evaluated once per system, rather than per-system and CUDA package set.)
```
This documentation belongs in `cuda-modules/_cuda`. For `all-packages.nix`' purposes, `_cuda` is a "top-level fixpoint used in `cudaPacakges`' internals"

SomeoneSerge commented on pkgs/top-level/cuda-packages.nix:
```diff
@@ -22,10 +22,11 @@
 # I've (@connorbaker) attempted to do that, though I'm unsure of how this will interact with overrides.
 {
   config,
+  cudaData,
+  cudaLib,
```
```suggestion
  _cuda,
```

ConnorBaker commented:
Will make your suggested changes when I get a chance (including stripping the `cuda` prefix from `cudaFixups` and friends).

SomeoneSerge commented:
I can push them as fixups for autosquash-ing to speed things up, shall I?

On May 22, 2025 3:55:36 PM UTC, Connor Baker ***@***.***> wrote:
>ConnorBaker left a comment (NixOS/nixpkgs#406531)
>
>Will make your suggested changes when I get a chance (including stripping the `cuda` prefix from `cudaFixups` and friends).
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/406531#issuecomment-2901760457
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

SomeoneSerge reviewed (approved):
@ConnorBaker I didn't want to overwrite your commits, leaving the autosquash to you. IMO this is ready to be merged to master, but let's have another look-over and maybe a nixpkgs-review before backporting
